common/src/main/java/org/apache/sedona/common/S2Geography/GeographyIndex.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.sedona.common.S2Geography;
+
+import com.google.common.geometry.S2CellId;
+import com.google.common.geometry.S2Iterator;
+import com.google.common.geometry.S2ShapeIndex;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Set;
+
+public class GeographyIndex {
+  private final S2ShapeIndex index;
+  private int[] values;
+
+  public GeographyIndex() {
+    this(new S2ShapeIndex.Options());
+  }
+
+  public GeographyIndex(S2ShapeIndex.Options options) {
+    this.index = new S2ShapeIndex(options);
+    this.values = new int[0];
+  }
+
+  public void add(S2Geography geog, int value) {
+    // Determine total slots needed in the array
+    int currentLength = (values == null ? 0 : values.length);
+    int needed = currentLength + geog.numShapes();
+    values = expand(values, needed);
```

Comment:
I suggest defining `values` as an `ArrayList`, so that it grows its space exponentially.

[The original C++ implementation](https://github.com/paleolimbot/s2geography/blob/main/src/s2geography/index.h#L30-L37) does not exhibit the exponential growth behavior and may incur lots of reallocation overhead, I think it could be a problem. @paleolimbot